### PR TITLE
persist: hook up decoding from only the structured arrow data

### PR DESCRIPTION
**Pull request by danhhz**
This is done in a way that  relies on the Codec data still being
populated (i.e. for the consolidate optimization above it). Eventually
we'll have to rewrite this path to work entirely without Codec data, but
in the meantime, putting in here allows us to see the performance impact
of decoding entirely from arrow instead of Codec.

Plus, it'll likely be easier to port all the logic here to work solely
on arrow data once we finish migrating things like the
ConsolidatingIter.

Touches https://github.com/MaterializeInc/materialize/issues/24830

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by danhhz** [comment]
Will run a full nightlies before merging (and fix this lint failure) but otherwise should be ready for review!

File: src/persist-client/src/fetch.rs [hunk 1]
```diff
@@ -744,15 +748,16 @@ impl<K: Codec, V: Codec, T: Timestamp + Lattice + Codec64, D> FetchedPart<K, V,
...
-        let structured_part = match (&part.part.updates, structured_part_audit) {
+        let should_downcast = match part_decode_format {
+            PartDecodeFormat::Row {
+                validate_structured,
+            } => validate_structured,
+            PartDecodeFormat::Arrow => true,
+        };
+        let structured_part = match (&part.part.updates, should_downcast) {
             // Only downcast and create decoders if we have structured data AND
             // an audit of the data is requested.
```
Comment (bkirwi) [inline src/persist-client/src/fetch.rs:759]: Comment needs an update!

**Review by bkirwi — APPROVED** [review]
It's good!

File: src/persist-client/src/fetch.rs (hunk 1 above)
Comment (danhhz) [reply src/persist-client/src/fetch.rs:759]: Done!

**Comment by danhhz** [comment]
TFTR!

**Pull request merged**